### PR TITLE
X3D parser: first cut, define/use and SD card, replaces PR #641

### PR DESCRIPTION
**Pull request by NolaDonato**
Combining all X3D parser work into a single commit.
define/use supported for most types.
Reading textures from SD card and subdirectories supported.

GearVRf-DCO-1.0-Signed-off-by: nola.donato@samsung.com Nola Donato and m1.williams@partner.samsung.com Mitch Williams

### CODE REVIEW

**Comment by liaxim** [comment]
Fyi, needs rebase

File: GVRf/Framework/framework/src/main/java/org/gearvrf/GVRAssetLoader.java [hunk 1]
```diff
@@ -502,48 +510,12 @@ public GVRModelSceneObject loadModel(String filePath, GVRScene scene) throws IOE
...
-     * @param scene
-     *            If present, this asset loader will wait until all of the textures have been
-     *            loaded and then adds the model to the scene.
-     *            
-     * @return A {@link GVRModelSceneObject} that contains the meshes with textures and bones
-     * and animations.
-     * @throws IOException 
-     *
-     */
-    public GVRModelSceneObject loadModel(String filePath, GVRResourceVolume.VolumeType volumeType, GVRScene scene) throws IOException
```
Comment (rahul27) [inline GVRf/Framework/framework/src/main/java/org/gearvrf/GVRAssetLoader.java]: Removing this call might break samples using this api. Better to deprecate it,

File: GVRf/Framework/framework/src/main/java/org/gearvrf/GVRAssetLoader.java [hunk 2]
```diff
@@ -949,11 +924,53 @@ private GVRSceneObject createSceneObject(
...
-                                               new Object[] { mContext, "X3D not supported yet", assetRequest.getFileName() });
-          return null;
+                    throws IOException {
+        GVRModelSceneObject root = new GVRModelSceneObject(mContext);
+        GVRResourceVolume volume = assetRequest.getVolume();
+        InputStream inputStream = null;
+        String fileName = assetRequest.getBaseName();
+        GVRAndroidResource resource = volume.openResource(fileName);
+
+        org.gearvrf.x3d.X3Dobject x3dObject = new org.gearvrf.x3d.X3Dobject(assetRequest, root);
```
Comment (rahul27) [inline GVRf/Framework/framework/src/main/java/org/gearvrf/GVRAssetLoader.java]: Use imports, makes the code cleaner.

File: GVRf/Framework/framework/src/main/java/org/gearvrf/x3d/DefinedItem.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  public void setGVRMaterial (GVRMaterial gvrMaterial) {
+    this.gvrMaterial = gvrMaterial;
+  }
+
+  public GVRMaterial getGVRMaterial() {
+    return this.gvrMaterial;
+  }
+
+  /*
+  public void setShaderSettingsMaterial(ShaderSettings shaderSettingsMaterial) {
```
Comment (rahul27) [inline GVRf/Framework/framework/src/main/java/org/gearvrf/x3d/DefinedItem.java]: remove unwanted comments.

File: GVRf/Framework/framework/src/main/java/org/gearvrf/x3d/DefinedItem.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ * Another item using the USE="...." parameter will search the array list for the 
+ * matching item.
+ */
+
+import org.gearvrf.GVRMesh;
+import org.gearvrf.GVRRenderData;
+import org.gearvrf.GVRSceneObject;
+import org.gearvrf.GVRTexture;
+
+public class DefinedItem {
```
Comment (rahul27) [inline GVRf/Framework/framework/src/main/java/org/gearvrf/x3d/DefinedItem.java]: Does this class need to be public? If so then it is missing javadocs.

File: GVRf/Framework/framework/src/main/java/org/gearvrf/GVRAssetLoader.java (hunk 1 above)
Comment (liaxim) [inline GVRf/Framework/framework/src/main/java/org/gearvrf/GVRAssetLoader.java]: I think it is time to start cleaning up all these compatibility loadModels

**Comment by rahul27** [comment]
Here are general comments:

- Public classes need javadocs
- The formatting seems off for a few files. Needs to be fixed
- Unwanted comments must be removed
- Avoid using System.out.print()
- Some of the java naming conventions do not follow the standard: http://java.about.com/od/understandingdatatypes/a/declaringvars.htm

**Comment by mwitchwilliams** [comment]
Formatted code, added javadocs comments and set some methods in X3Dobject private (or removed in some cases).  System.out.print() should be removed too.

File: GVRf/Framework/framework/src/main/java/org/gearvrf/x3d/DefinedItem.java (hunk 4 above)
Comment (mwitchwilliams) [inline GVRf/Framework/framework/src/main/java/org/gearvrf/x3d/DefinedItem.java]: Added javadoc comments.  As public, GearVRf's X3D implementation enables developers to access any DEFined object, similar to HTML5's DOM's (Document Object Model) getElementsByTagName("_def_name_here_"); method.  Currently, only GVRSceneObject allows access by name, but in X3D the lights, 3d meshes, texture shoul all be accessed by name.

File: GVRf/Framework/framework/src/main/java/org/gearvrf/x3d/DefinedItem.java (hunk 3 above)
Comment (mwitchwilliams) [inline GVRf/Framework/framework/src/main/java/org/gearvrf/x3d/DefinedItem.java]: done.

**Comment by liaxim** [comment]
Fyi, needs rebase

**Comment by mwitchwilliams** [comment]
rebase done

**Comment by liaxim** [comment]
Will merge after the rebase.

**Pull request closed without merging**